### ISSUE TITLE
data race on ipfs/libp2p

### ISSUE THREAD

**Issue by p4u** [vocdoni/vocdoni-node#104]
```
WARNING: DATA RACE
Read at 0x00c05872b5e0 by goroutine 738:
  github.com/ipfs/go-bitswap/internal/decision.(*blockstoreManager).addJob()
      /go/pkg/mod/github.com/ipfs/go-bitswap@v0.3.3/internal/decision/blockstoremanager.go:58 +0x84
  github.com/ipfs/go-bitswap/internal/decision.(*blockstoreManager).jobPerKey()
      /go/pkg/mod/github.com/ipfs/go-bitswap@v0.3.3/internal/decision/blockstoremanager.go:115 +0x1ef
  github.com/ipfs/go-bitswap/internal/decision.(*blockstoreManager).getBlockSizes()
      /go/pkg/mod/github.com/ipfs/go-bitswap@v0.3.3/internal/decision/blockstoremanager.go:72 +0x1bc
  github.com/ipfs/go-bitswap/internal/decision.(*Engine).MessageReceived()
      /go/pkg/mod/github.com/ipfs/go-bitswap@v0.3.3/internal/decision/engine.go:453 +0x36d
  github.com/ipfs/go-bitswap.(*Bitswap).ReceiveMessage()
      /go/pkg/mod/github.com/ipfs/go-bitswap@v0.3.3/bitswap.go:444 +0x169
  github.com/ipfs/go-bitswap/network.(*impl).handleNewStream()
      /go/pkg/mod/github.com/ipfs/go-bitswap@v0.3.3/network/ipfs_impl.go:402 +0x562
  github.com/ipfs/go-bitswap/network.(*impl).handleNewStream-fm()
      /go/pkg/mod/github.com/ipfs/go-bitswap@v0.3.3/network/ipfs_impl.go:378 +0x55
  github.com/libp2p/go-libp2p/p2p/host/basic.(*BasicHost).SetStreamHandler.func1()
      /go/pkg/mod/github.com/libp2p/go-libp2p@v0.13.0/p2p/host/basic/basic_host.go:566 +0xb4

Previous write at 0x00c05872b5e0 by goroutine 52:
  github.com/ipfs/go-bitswap/internal/decision.(*blockstoreManager).start()
      /go/pkg/mod/github.com/ipfs/go-bitswap@v0.3.3/internal/decision/blockstoremanager.go:33 +0x44
  github.com/ipfs/go-bitswap/internal/decision.(*Engine).StartWorkers()
      /go/pkg/mod/github.com/ipfs/go-bitswap@v0.3.3/internal/decision/engine.go:235 +0x6a
  github.com/ipfs/go-bitswap.New()
      /go/pkg/mod/github.com/ipfs/go-bitswap@v0.3.3/bitswap.go:219 +0x18b4
  github.com/ipfs/go-ipfs/core/node.OnlineExchange.func1()
      /go/pkg/mod/github.com/ipfs/go-ipfs@v0.7.1-0.20210129042248-884a5aebd748/core/node/core.go:94 +0x1e4
  runtime.call128()
      /usr/local/go/src/runtime/asm_amd64.s:542 +0x58
  reflect.Value.Call()
      /usr/local/go/src/reflect/value.go:337 +0xd8
  go.uber.org/dig.defaultInvoker()
      /go/pkg/mod/go.uber.org/dig@v1.10.0/dig.go:284 +0x7a
  go.uber.org/dig.(*node).Call()
      /go/pkg/mod/go.uber.org/dig@v1.10.0/dig.go:710 +0x2d0
  go.uber.org/dig.paramSingle.Build()
      /go/pkg/mod/go.uber.org/dig@v1.10.0/param.go:245 +0x39a
  go.uber.org/dig.(*paramSingle).Build()
      <autogenerated>:1 +0xe7
  go.uber.org/dig.paramList.BuildList()
      /go/pkg/mod/go.uber.org/dig@v1.10.0/param.go:201 +0xf7
  go.uber.org/dig.(*node).Call()
      /go/pkg/mod/go.uber.org/dig@v1.10.0/dig.go:701 +0x155
  go.uber.org/dig.paramSingle.Build()
      /go/pkg/mod/go.uber.org/dig@v1.10.0/param.go:245 +0x39a
  go.uber.org/dig.(*paramSingle).Build()
      <autogenerated>:1 +0xe7
  go.uber.org/dig.paramList.BuildList()
      /go/pkg/mod/go.uber.org/dig@v1.10.0/param.go:201 +0xf7
  go.uber.org/dig.(*node).Call()
      /go/pkg/mod/go.uber.org/dig@v1.10.0/dig.go:701 +0x155
  go.uber.org/dig.paramSingle.Build()
      /go/pkg/mod/go.uber.org/dig@v1.10.0/param.go:245 +0x39a
  go.uber.org/dig.(*paramSingle).Build()
      <autogenerated>:1 +0xe7
  go.uber.org/dig.paramList.BuildList()
      /go/pkg/mod/go.uber.org/dig@v1.10.0/param.go:201 +0xf7
  go.uber.org/dig.(*node).Call()
      /go/pkg/mod/go.uber.org/dig@v1.10.0/dig.go:701 +0x155
  go.uber.org/dig.paramSingle.Build()
      /go/pkg/mod/go.uber.org/dig@v1.10.0/param.go:245 +0x39a
  go.uber.org/dig.(*paramSingle).Build()
      <autogenerated>:1 +0xe7
  go.uber.org/dig.paramObjectField.Build()
      /go/pkg/mod/go.uber.org/dig@v1.10.0/param.go:384 +0x1c3
  go.uber.org/dig.paramObject.Build()
      /go/pkg/mod/go.uber.org/dig@v1.10.0/param.go:311 +0x160
  go.uber.org/dig.(*paramObject).Build()
      <autogenerated>:1 +0xe7
  go.uber.org/dig.paramList.BuildList()
      /go/pkg/mod/go.uber.org/dig@v1.10.0/param.go:201 +0xf7
  go.uber.org/dig.(*Container).Invoke()
      /go/pkg/mod/go.uber.org/dig@v1.10.0/dig.go:432 +0x384
  go.uber.org/fx.(*App).executeInvokes()
      /go/pkg/mod/go.uber.org/fx@v1.13.1/app.go:692 +0x5f8
  go.uber.org/fx.New()
      /go/pkg/mod/go.uber.org/fx@v1.13.1/app.go:471 +0xb5b
  github.com/ipfs/go-ipfs/core.NewNode()
      /go/pkg/mod/github.com/ipfs/go-ipfs@v0.7.1-0.20210129042248-884a5aebd748/core/builder.go:41 +0x2ef
  go.vocdoni.io/dvote/ipfs.StartNode()
      /__w/vocdoni-node/vocdoni-node/ipfs/ipfs.go:72 +0x2cd
  github.com/syndtr/goleveldb/leveldb.(*DB).checkAndCleanFiles()
      /go/pkg/mod/github.com/syndtr/goleveldb@v1.0.1-0.20200815110645-5c35d600f0ca/leveldb/db_util.go:52 +0x301
  github.com/syndtr/goleveldb/leveldb.openDB()
      /go/pkg/mod/github.com/syndtr/goleveldb@v1.0.1-0.20200815110645-5c35d600f0ca/leveldb/db.go:136 +0xa3e
  github.com/syndtr/goleveldb/leveldb/storage.(*fileStorage).List()
      /go/pkg/mod/github.com/syndtr/goleveldb@v1.0.1-0.20200815110645-5c35d600f0ca/leveldb/storage/file_storage.go:458 +0x319
  fmt.Fscanf()
      /usr/local/go/src/fmt/scan.go:143 +0xee
  fmt.Sscanf()
      /usr/local/go/src/fmt/scan.go:114 +0x191
  github.com/syndtr/goleveldb/leveldb/storage.fsParseName()
      /go/pkg/mod/github.com/syndtr/goleveldb@v1.0.1-0.20200815110645-5c35d600f0ca/leveldb/storage/file_storage.go:643 +0xa6
  github.com/syndtr/goleveldb/leveldb/storage.(*fileStorage).List()
      /go/pkg/mod/github.com/syndtr/goleveldb@v1.0.1-0.20200815110645-5c35d600f0ca/leveldb/storage/file_storage.go:458 +0x319
  fmt.Fscanf()
      /usr/local/go/src/fmt/scan.go:143 +0xee
  fmt.Sscanf()
      /usr/local/go/src/fmt/scan.go:114 +0x284
  github.com/syndtr/goleveldb/leveldb/storage.fsParseName()
      /go/pkg/mod/github.com/syndtr/goleveldb@v1.0.1-0.20200815110645-5c35d600f0ca/leveldb/storage/file_storage.go:657 +0x19e
  fmt.(*ss).doScanf()
      /usr/local/go/src/fmt/scan.go:1230 +0x3c5
  fmt.Fscanf()
      /usr/local/go/src/fmt/scan.go:143 +0xee
  fmt.Sscanf()
      /usr/local/go/src/fmt/scan.go:114 +0x191
  github.com/syndtr/goleveldb/leveldb/storage.fsParseName()
      /go/pkg/mod/github.com/syndtr/goleveldb@v1.0.1-0.20200815110645-5c35d600f0ca/leveldb/storage/file_storage.go:643 +0xa6
  github.com/syndtr/goleveldb/leveldb/storage.(*fileStorage).List()
      /go/pkg/mod/github.com/syndtr/goleveldb@v1.0.1-0.20200815110645-5c35d600f0ca/leveldb/storage/file_storage.go:458 +0x319
  fmt.Fscanf()
      /usr/local/go/src/fmt/scan.go:143 +0xee
  fmt.Sscanf()
      /usr/local/go/src/fmt/scan.go:114 +0x191
  github.com/syndtr/goleveldb/leveldb/storage.fsParseName()
      /go/pkg/mod/github.com/syndtr/goleveldb@v1.0.1-0.20200815110645-5c35d600f0ca/leveldb/storage/file_storage.go:643 +0xa6
  github.com/syndtr/goleveldb/leveldb/storage.(*fileStorage).List()
      /go/pkg/mod/github.com/syndtr/goleveldb@v1.0.1-0.20200815110645-5c35d600f0ca/leveldb/storage/file_storage.go:458 +0x319
  github.com/syndtr/goleveldb/leveldb.(*DB).recoverJournal()
      /go/pkg/mod/github.com/syndtr/goleveldb@v1.0.1-0.20200815110645-5c35d600f0ca/leveldb/db.go:482 +0xc8
  github.com/syndtr/goleveldb/leveldb.openDB()
      /go/pkg/mod/github.com/syndtr/goleveldb@v1.0.1-0.20200815110645-5c35d600f0ca/leveldb/db.go:131 +0xa1a
  github.com/syndtr/goleveldb/leveldb.Open()
      /go/pkg/mod/github.com/syndtr/goleveldb@v1.0.1-0.20200815110645-5c35d600f0ca/leveldb/db.go:203 +0x247
  fmt.Fscanf()
      /usr/local/go/src/fmt/scan.go:143 +0xee
  fmt.Sscanf()
      /usr/local/go/src/fmt/scan.go:114 +0x191
  github.com/syndtr/goleveldb/leveldb/storage.fsParseName()
      /go/pkg/mod/github.com/syndtr/goleveldb@v1.0.1-0.20200815110645-5c35d600f0ca/leveldb/storage/file_storage.go:643 +0xa6
  github.com/syndtr/goleveldb/leveldb/storage.(*fileStorage).List()
      /go/pkg/mod/github.com/syndtr/goleveldb@v1.0.1-0.20200815110645-5c35d600f0ca/leveldb/storage/file_storage.go:458 +0x319
  github.com/syndtr/goleveldb/leveldb.(*session).recover.func1()
      /go/pkg/mod/github.com/syndtr/goleveldb@v1.0.1-0.20200815110645-5c35d600f0ca/leveldb/session.go:134 +0x103
  github.com/syndtr/goleveldb/leveldb.(*session).recover()
      /go/pkg/mod/github.com/syndtr/goleveldb@v1.0.1-0.20200815110645-5c35d600f0ca/leveldb/session.go:142 +0x1952
  github.com/syndtr/goleveldb/leveldb.Open()
      /go/pkg/mod/github.com/syndtr/goleveldb@v1.0.1-0.20200815110645-5c35d600f0ca/leveldb/db.go:189 +0x109
  github.com/syndtr/goleveldb/leveldb.OpenFile()
      /go/pkg/mod/github.com/syndtr/goleveldb@v1.0.1-0.20200815110645-5c35d600f0ca/leveldb/db.go:225 +0xae
  github.com/ipfs/go-ds-leveldb.NewDatastore()
      /go/pkg/mod/github.com/ipfs/go-ds-leveldb@v0.4.2/datastore.go:46 +0x3d9
  github.com/ipfs/go-ipfs/plugin/plugins/levelds.(*datastoreConfig).Create()
      /go/pkg/mod/github.com/ipfs/go-ipfs@v0.7.1-0.20210129042248-884a5aebd748/plugin/plugins/levelds/levelds.go:85 +0xfd
  github.com/ipfs/go-ipfs/repo/fsrepo.measureDatastoreConfig.Create()
      /go/pkg/mod/github.com/ipfs/go-ipfs@v0.7.1-0.20210129042248-884a5aebd748/repo/fsrepo/datastores.go:243 +0x5c
  github.com/ipfs/go-ipfs/repo/fsrepo.(*measureDatastoreConfig).Create()
      <autogenerated>:1 +0xac
  github.com/ipfs/go-ipfs/repo/fsrepo.(*mountDatastoreConfig).Create()
      /go/pkg/mod/github.com/ipfs/go-ipfs@v0.7.1-0.20210129042248-884a5aebd748/repo/fsrepo/datastores.go:154 +0x13e
  github.com/ipfs/go-ipfs/repo/fsrepo.(*FSRepo).openDatastore()
      /go/pkg/mod/github.com/ipfs/go-ipfs@v0.7.1-0.20210129042248-884a5aebd748/repo/fsrepo/fsrepo.go:440 +0x499
  github.com/ipfs/go-ipfs/repo/fsrepo.open()
      /go/pkg/mod/github.com/ipfs/go-ipfs@v0.7.1-0.20210129042248-884a5aebd748/repo/fsrepo/fsrepo.go:169 +0x44c
  github.com/ipfs/go-ipfs/repo/fsrepo.Open.func1()
      /go/pkg/mod/github.com/ipfs/go-ipfs@v0.7.1-0.20210129042248-884a5aebd748/repo/fsrepo/fsrepo.go:113 +0x4e
  github.com/ipfs/go-ipfs/repo.(*OnlyOne).Open()
      /go/pkg/mod/github.com/ipfs/go-ipfs@v0.7.1-0.20210129042248-884a5aebd748/repo/onlyone.go:35 +0x207
  github.com/ipfs/go-ipfs/repo/fsrepo.Open()
      /go/pkg/mod/github.com/ipfs/go-ipfs@v0.7.1-0.20210129042248-884a5aebd748/repo/fsrepo/fsrepo.go:115 +0x9c
  go.vocdoni.io/dvote/ipfs.StartNode()
      /__w/vocdoni-node/vocdoni-node/ipfs/ipfs.go:53 +0x1b0
  go.vocdoni.io/dvote/data.(*IPFSHandle).Init()
      /__w/vocdoni-node/vocdoni-node/data/ipfs.go:60 +0x1a4
  go.vocdoni.io/dvote/test/testcommon.(*DvoteAPIServer).Start()
      /__w/vocdoni-node/vocdoni-node/test/testcommon/apis.go:67 +0x65e
  go.vocdoni.io/dvote/test.TestCensus()
      /__w/vocdoni-node/vocdoni-node/test/census_test.go:59 +0x148
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1123 +0x202

Goroutine 738 (running) created at:
  github.com/libp2p/go-libp2p/p2p/host/basic.(*BasicHost).newStreamHandler()
      /go/pkg/mod/github.com/libp2p/go-libp2p@v0.13.0/p2p/host/basic/basic_host.go:416 +0x7a4
  github.com/libp2p/go-libp2p/p2p/host/basic.(*BasicHost).newStreamHandler-fm()
      /go/pkg/mod/github.com/libp2p/go-libp2p@v0.13.0/p2p/host/basic/basic_host.go:373 +0x55
  github.com/libp2p/go-libp2p-swarm.(*Conn).start.func1.1()
      /go/pkg/mod/github.com/libp2p/go-libp2p-swarm@v0.4.0/swarm_conn.go:125 +0x11a

Goroutine 52 (running) created at:
  testing.(*T).Run()
      /usr/local/go/src/testing/testing.go:1168 +0x5bb
  testing.runTests.func1()
      /usr/local/go/src/testing/testing.go:1439 +0xa6
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1123 +0x202
  testing.runTests()
      /usr/local/go/src/testing/testing.go:1437 +0x612
  testing.(*M).Run()
      /usr/local/go/src/testing/testing.go:1345 +0x3b3
  main.main()
      _testmain.go:113 +0x356
 ```

**Comment by mvdan** [comment]
@p4u what test triggered this? that would help me reproduce and debug the data race

**Comment by mvdan** [comment]
Ah, I see from the log that it was probably TestCensus.

**Comment by mvdan** [comment]
I tried to minify this a bit, but I just ended up finding that combining ipfs/repo/fsrepo with ipfs/core seems to result in a data race. I filed https://github.com/ipfs/go-ipfs/issues/7947 for it.

**Comment by p4u** [comment]
What's the status of this @mvdan?

**Comment by mvdan** [comment]
I was waiting on the two fixes to get merged, and now that's happened. I'll send a PR to update the libraries. We'll have to use master versions of those two libs for a little while, as they won't do an ipfs release with these upstream fixes for months.
